frame/dvm/ethereum/src/lib.rs

```diff
@@ -0,0 +1,464 @@
+// Copyright 2017-2020 Parity Technologies (UK) Ltd.
+// This file is part of Frontier.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Ethereum pallet
+//!
+//! The Ethereum pallet works together with EVM pallet to provide full emulation
+//! for Ethereum block processing.
+
+// Ensure we're `no_std` when compiling for Wasm.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::Encode;
+use dvm_consensus_primitives::{ConsensusLog, FRONTIER_ENGINE_ID};
+use ethereum_types::{Bloom, H160, H256, H64, U256};
+use evm::{ExitError, ExitFatal, ExitReason, ExitRevert};
+use frame_support::{
+	decl_error, decl_event, decl_module, decl_storage, traits::FindAuthor, traits::Get,
+};
+use frame_system::ensure_none;
+use sha3::{Digest, Keccak256};
+use sp_runtime::{
+	generic::DigestItem,
+	traits::UniqueSaturatedInto,
+	transaction_validity::{
+		InvalidTransaction, TransactionSource, TransactionValidity, ValidTransaction,
+	},
+	DispatchResult,
+};
+use sp_std::prelude::*;
+
+pub use dvm_rpc_primitives::TransactionStatus;
+pub use ethereum::{Block, Log, Receipt, Transaction, TransactionAction};
+use frame_support::traits::Currency;
+use pallet_evm::AddressMapping;
+
+#[cfg(all(feature = "std", test))]
+mod tests;
+
+#[cfg(all(feature = "std", test))]
+mod mock;
+// Precomile contract for dvm
+pub mod precompiles;
+
+/// A type alias for the balance type from this pallet's point of view.
+pub type BalanceOf<T> = <T as darwinia_balances::Trait>::Balance;
+type RingInstance = darwinia_balances::Instance0;
+
+/// Trait for Ethereum pallet.
+pub trait Trait:
+	frame_system::Trait<Hash = H256>
+	+ darwinia_balances::Trait<RingInstance>
+	+ pallet_timestamp::Trait
+	+ pallet_evm::Trait
+{
+	/// The overarching event type.
+	type Event: From<Event> + Into<<Self as frame_system::Trait>::Event>;
+	/// Find author for Ethereum.
+	type FindAuthor: FindAuthor<H160>;
+	type AddressMapping: AddressMapping<Self::AccountId>;
+	type RingCurrency: Currency<Self::AccountId>;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Ethereum {
```

Comment:
Do we need a different module storage prefix?

Reply:
Maybe not enough reason to do these changes at present.